src/expr/src/scalar/func.rs

```diff
@@ -1321,6 +1321,16 @@ fn neg_decimal<'a>(a: Datum<'a>) -> Datum<'a> {
     Datum::from(-a.unwrap_decimal())
 }
 
+fn neg_apd<'a>(a: Datum<'a>) -> Datum<'a> {
+    let mut a = a.unwrap_apd();
+    // 0 and NaN can be negated in the underlying library, but must not be
+    // negated in SQL.
+    if !(a.0.is_zero() || a.0.is_nan()) {
+        apd::cx_datum().neg(&mut a.0);
+    }
```

Comment:
I'm a little worried about trying to enforce this adhoc. We might want to write a function (`munge_apd`) that gets called after every op on an APD that handles this.

Reply:
Ah, ok––I had thought of that but it seemed a little wasteful on CPU cycles for things we would need to test for in the first place. If you prefer `munge_apd`, though, glad to fix that up here.